### PR TITLE
[POAE7-2417] Support NotNull Ops for Arrow Format and some fix

### PR DESCRIPTION
**Pull request by Deegue**
### What changes were proposed in this pull request?
Some fixes and improvements to make it an easier way to use new Arrow Format including data generator and checker.

1. Fix empty result arrow batch core dump.
2. Fix compile error when using both data generator.
3. Fix and use `checkArrowEq` in `assertQueryArrow`.
4. Fix arrow format in `SetUp` of `CiderTestBase`.
5. Add and enable tests for null and not null.

### Why are the changes needed?
fix & improvements.

### Does this PR introduce _any_ user-facing change?
No.

### How was this patch tested?
UTs.

### CODE REVIEW

**Comment by Deegue** [comment]
This PR should be moved on(review & retest) after https://github.com/intel/BDTK/pull/90 is merged.

File: cider/tests/functionality/CiderFilterTest.cpp [hunk 1]
```diff
@@ -20,265 +20,157 @@
...
+         CREATE_SUBSTRAIT_TYPE(Fp64)},
+        {1, 2, 2, 2, 3, 3},
+        GeneratePattern::Random);
+    input_ = {std::make_shared<CiderBatch>(
+        *(new CiderBatch(schema, array, std::make_shared<CiderDefaultAllocator>())))};
   }
 };
 
 TEST_F(CiderProjectAllTestBase, filterProjectAllTest) {
-  assertQuery("SELECT * FROM test");
```
Comment (yma11) [inline cider/tests/functionality/CiderFilterTest.cpp:139]: Test cases for old data format need be kept until we finished migration.

File: cider/tests/functionality/CiderFilterTest.cpp [hunk 2]
```diff
@@ -20,265 +20,157 @@
...
-         CREATE_SUBSTRAIT_TYPE(Fp64),
-         CREATE_SUBSTRAIT_TYPE(Date),
-         CREATE_SUBSTRAIT_TYPE(Bool)},
-        {1, 2, 2, 2, 3, 3, 4, 2},
-        GeneratePattern::Random))};
+         CREATE_SUBSTRAIT_TYPE(Fp64)},
+        {1, 2, 2, 2, 3, 3},
+        GeneratePattern::Random);
+    input_ = {std::make_shared<CiderBatch>(
+        *(new CiderBatch(schema, array, std::make_shared<CiderDefaultAllocator>())))};
```
Comment (yma11) [inline cider/tests/functionality/CiderFilterTest.cpp:152]: The usage seems weird. Can we provide a similar API by doing this things inside `CiderTestBase`?

File: cider/tests/utils/DuckDbQueryRunner.cpp [hunk 3]
```diff
@@ -532,7 +532,25 @@ DuckDbResultConvertor::fetchDataToArrowFormattedCiderBatch(
...
+        arrow_array->release = CiderBatchUtils::ciderEmptyArrowArrayReleaser;
+
+        arrow_schema->format = "+s";
+        arrow_schema->dictionary = nullptr;
+        arrow_schema->n_children = 0;
+        arrow_schema->children = (ArrowSchema**)std::malloc(0);
+        arrow_schema->release = CiderBatchUtils::ciderEmptyArrowSchemaReleaser;
+
+        batch_res.push_back(std::make_shared<CiderBatch>(*(new CiderBatch(
+            arrow_schema, arrow_array, std::make_shared<CiderDefaultAllocator>()))));
```
Comment (yma11) [inline cider/tests/utils/DuckDbQueryRunner.cpp:553]: I remember DuckDb result convertor should be already implemented by this [PR](https://github.com/intel/BDTK/pull/90). please check.

Comment (Deegue) [reply cider/tests/utils/DuckDbQueryRunner.cpp:553]: Here we should handle empty batch get from DuckDB. If we return a CiderBatch leaving schema and array nullptr, many methods (like getLength etc.) will core. And I think the PR should be https://github.com/intel/BDTK/pull/67.

File: cider/tests/functionality/CiderFilterTest.cpp (hunk 2 above)
Comment (Deegue) [reply cider/tests/functionality/CiderFilterTest.cpp]: Moved to `CiderTestBase`.

File: cider/tests/functionality/CiderFilterTest.cpp (hunk 1 above)
Comment (Deegue) [reply cider/tests/functionality/CiderFilterTest.cpp]: Reverted and keep all old test cases.

File: cider/tests/utils/DuckDbQueryRunner.cpp (hunk 3 above)
Comment (yma11) [reply cider/tests/utils/DuckDbQueryRunner.cpp:553]: Got U. Just make sure there is no duplicated change about this part.

File: cider/tests/utils/DuckDbQueryRunner.cpp [hunk 4]
```diff
@@ -532,7 +532,25 @@ DuckDbResultConvertor::fetchDataToArrowFormattedCiderBatch(
     auto chunk = result->Fetch();
     if (!chunk || (chunk->size() == 0)) {
       if (batch_res.empty()) {
-        batch_res.push_back(std::make_shared<CiderBatch>());
+        auto arrow_array = CiderBatchUtils::allocateArrowArray();
+        auto arrow_schema = CiderBatchUtils::allocateArrowSchema();
+
+        arrow_array->length = 0;
```
Comment (jikunshang) [inline cider/tests/utils/DuckDbQueryRunner.cpp:538]: these are same to `CiderBatchUtils::allocateArrowArray`?

File: cider/tests/utils/DuckDbQueryRunner.cpp [hunk 5]
```diff
@@ -532,7 +532,25 @@ DuckDbResultConvertor::fetchDataToArrowFormattedCiderBatch(
...
+        arrow_array->n_children = 0;
+        arrow_array->buffers = nullptr;
+        arrow_array->n_buffers = 0;
+        arrow_array->private_data = nullptr;
+        arrow_array->children = (ArrowArray**)std::malloc(0);
```
Comment (jikunshang) [inline cider/tests/utils/DuckDbQueryRunner.cpp:543]: what does malloc(0) returns ? and we should avoid call malloc directly.

File: cider/tests/utils/CiderTestBase.h [hunk 6]
```diff
@@ -65,6 +73,8 @@ class CiderTestBase : public testing::Test {
   std::string table_name_;
   std::string create_ddl_;
   std::vector<std::shared_ptr<CiderBatch>> input_;
+  ArrowArray* array = nullptr;
```
Comment (jikunshang) [inline cider/tests/utils/CiderTestBase.h:76]: `array_` & `schema_`

File: cider/tests/utils/CiderTestBase.cpp [hunk 7]
```diff
@@ -50,12 +50,26 @@ void CiderTestBase::prepareArrowBatch() {
...
                                      const std::string& json_file) {
   auto duck_res = duckDbQueryRunner_.runSql(sql);
-  auto duck_res_batch = DuckDbResultConvertor::fetchDataToCiderBatch(duck_res);
+  auto duck_res_batch =
+      DuckDbResultConvertor::fetchDataToArrowFormattedCiderBatch(duck_res);
+
+  auto cider_input = json_file.size() ? json_file : sql;
+  auto cider_res_batch = std::make_shared<CiderBatch>(
+      ciderQueryRunner_.runQueryOneBatch(cider_input, input_[0], true));
+  EXPECT_TRUE(CiderBatchChecker::checkArrowEq(duck_res_batch, cider_res_batch, false));
```
Comment (jikunshang) [inline cider/tests/utils/CiderTestBase.cpp:59]: so we can remove `checkArrowEqTemp`? and `convertToArrowRepresentation`

File: cider/tests/utils/CiderTestBase.cpp [hunk 8]
```diff
@@ -50,12 +50,26 @@ void CiderTestBase::prepareArrowBatch() {
...
+  auto duck_res = duckDbQueryRunner_.runSql(sql);
+  auto duck_res_batch =
+      DuckDbResultConvertor::fetchDataToArrowFormattedCiderBatch(duck_res);
 
   auto cider_input = json_file.size() ? json_file : sql;
   auto cider_res_batch = std::make_shared<CiderBatch>(
       ciderQueryRunner_.runQueryOneBatch(cider_input, input_[0], true));
-  EXPECT_TRUE(CiderBatchChecker::checkArrowEqTemp(duck_res_batch, cider_res_batch));
+
+  EXPECT_TRUE(CiderBatchChecker::checkEq(duck_res_batch, cider_res_batch, true));
```
Comment (YBRua) [inline cider/tests/utils/CiderTestBase.cpp:72]: I think it probably should be `EXPECT_TRUE(CiderBatchChecker::checkArrowEq(...))` here?

Comment (Deegue) [reply cider/tests/utils/CiderTestBase.cpp:72]: Yes, just the same as above, let me revise it.

File: cider/tests/utils/CiderTestBase.h (hunk 6 above)
Comment (haiweizh) [inline cider/tests/utils/CiderTestBase.h:76]: protected variable's name should been end with _

**Review by haiweizh — CHANGES_REQUESTED** [review]

File: cider/tests/utils/CiderTestBase.h [hunk 9]
```diff
@@ -34,7 +34,13 @@
 class CiderTestBase : public testing::Test {
  public:
   void SetUp() override {
-    duckDbQueryRunner_.createTableAndInsertData(table_name_, create_ddl_, input_);
+    if (array && schema) {
+      input_ = {std::make_shared<CiderBatch>(
```
[the remark this answers is not in the document]
Comment (Deegue) [reply cider/tests/utils/CiderTestBase.h:38]: It works well in local testing. Will it be any trouble if we set input_ here?

File: cider/tests/utils/DuckDbQueryRunner.cpp (hunk 4 above)
Comment (Deegue) [reply cider/tests/utils/DuckDbQueryRunner.cpp]: redundant code removed.

File: cider/tests/utils/DuckDbQueryRunner.cpp (hunk 5 above)
Comment (Deegue) [reply cider/tests/utils/DuckDbQueryRunner.cpp]: I think we will get a valid pointer by malloc(0) instead of nullptr. 

Went through all the places using `getChildrenArrayPtr()` and it looks okay if we just set it to nullptr.

File: cider/tests/utils/CiderTestBase.h (hunk 6 above)
Comment (Deegue) [reply cider/tests/utils/CiderTestBase.h]: done

File: cider/tests/utils/CiderTestBase.cpp (hunk 7 above)
Comment (Deegue) [reply cider/tests/utils/CiderTestBase.cpp:59]: yes, I think so

File: cider/tests/utils/CiderTestBase.h (hunk 6 above)
Comment (Deegue) [reply cider/tests/utils/CiderTestBase.h]: Get it, thanks~

File: cider/tests/utils/CiderTestBase.h (hunk 9 above)
[the remark this answers is not in the document]
Comment (jikunshang) [reply cider/tests/utils/CiderTestBase.h:38]: `auto ptr = std::make_shared<Type>(* new Type())`  
`auto ptr = std::make_shared<Type>()`
the later one is better, right?

File: cider/tests/utils/CiderTestBase.cpp (hunk 7 above)
Comment (jikunshang) [reply cider/tests/utils/CiderTestBase.cpp:59]: seems some types are not totally ready(like string). let's keep them for a while.

File: cider/tests/functionality/CiderFilterTest.cpp [hunk 10]
```diff
@@ -276,9 +261,245 @@ TEST_F(CiderFilterSequenceTestBase, multiColEqualTest) {
...
+      true);
+}
+
+// Extends CiderTestBase and create a (99 rows, 4 types columns) table for filter test.
+class CiderFilterSequenceTestArrow : public CiderTestBase {
+ public:
+  CiderFilterSequenceTestArrow() {
+    table_name_ = "test";
+    create_ddl_ =
+        R"(CREATE TABLE test(col_1 INTEGER, col_2 BIGINT,
```
Comment (jikunshang) [inline cider/tests/functionality/CiderFilterTest.cpp:371]: Please add `NOT NULL` in create table command if you didn't generate null data. generated IR will be different.

File: cider/tests/functionality/CiderFilterTest.cpp [hunk 11]
```diff
@@ -398,117 +611,135 @@ TEST_F(CiderFilterSequenceTestArrow, ArrowMultiFilterWithAndTest) {
...
 }
 
-TEST_F(CiderFilterRandomTestBase, multiColRandomTest) {
-  assertQuery("SELECT col_1, col_5 FROM test WHERE col_1 < col_5");
-  assertQuery("SELECT col_2, col_6 FROM test WHERE col_2 < col_6");
-  assertQuery("SELECT col_3, col_7 FROM test WHERE col_3 <= col_7");
-  assertQuery("SELECT col_4, col_8 FROM test WHERE col_4 <= col_8");
-  assertQuery("SELECT col_1, col_5 FROM test WHERE col_1 <> col_5");
-  assertQuery(
+TEST_F(CiderFilterRandomTestArrow, multiColRandomTest) {
```
Comment (jikunshang) [inline cider/tests/functionality/CiderFilterTest.cpp:620]: should we keep original test cases?

Comment (Deegue) [reply cider/tests/functionality/CiderFilterTest.cpp:620]: yes, all original test cases are on top of this file.

File: cider/tests/functionality/CiderFilterTest.cpp (hunk 10 above)
Comment (Deegue) [reply cider/tests/functionality/CiderFilterTest.cpp:371]: done

File: cider/tests/functionality/CiderFilterTest.cpp (hunk 11 above)
Comment (jikunshang) [reply cider/tests/functionality/CiderFilterTest.cpp:619]: oh sorry. I didn't notice that. Can we re-org to reduce code diff?

Comment (Deegue) [reply cider/tests/functionality/CiderFilterTest.cpp:619]: never mind, original test cases interspersing with the new ones will reduce code diff, but we have to go into each case to find and remove the original ones when finished arrow migration.

**Review by jikunshang — APPROVED** [review]

**Review by yma11 — APPROVED** [review]
